protocol.md

```diff
@@ -726,6 +727,141 @@ Upload-Length: 11
 Upload-Concat: final;/files/a /files/b
 ```
 
+### Challenge
+
+With this extension, Clients can authenticate follow-up requests to a [Creation](#creation) or [Creation With Upload](#creation-with-upload) 
+`POST` request using a high-entropy cryptographic random shared secret - and challenges
+based thereon.
+
+If the Server supports this extension, it MUST add `challenge` to the `Tus-Extension` header.
+
+#### Headers
+
+##### Upload-Secret
+
+The `Upload-Secret` MUST be a high-entropy cryptographic random string, consisting
+of 48 to 256 characters from the [base64 alphabet](https://tools.ietf.org/html/rfc4648#section-4).
+That's 36 to 192 bytes encoded as base64.
+
+The `Upload-Secret` header MUST NOT be included with anything but `POST` requests
+that create a new upload resource. Servers receiving an `Upload-Secret` header with
+any other request MUST respond with a `400 Bad Request` status.
+
+##### Upload-Challenge
+
+If an `Upload-Secret` header was sent with the `POST` request that created
+the upload resource, Clients MUST include an `Upload-Challenge` header with every
+subsequent request targeting that upload resource.
+
+The value of `Upload-Challenge` is defined as follows:
+
+```
+Upload-Challenge = SHA256([Upload-Secret] + SHA256(`[HTTP Method]` + [Upload-Offset] + [Upload-Secret] + [Content-Length]))
+```
+
+The *effective* HTTP method of the request that the `Upload-Challenge` is sent with MUST be
+used in place of `HTTP Method`. If `X-HTTP-Method-Override` is used in the same request,
+`HTTP method` needs to be replaced with the value of `X-HTTP-Method-Override`.
+
+Where `Upload-Offset` and `Content-Length` headers are used in the same request, their
+values MUST be used as the respective `Upload-Offset` and `Content-Length` values. A
+value of `0` MUST be used for `Upload-Offset` and `Content-Length` where no headers
+of the same name are sent with the request.
+
+Servers receiving a `HEAD` or `PATCH` request with a missing or invalid `Upload-Challenge` value
+MUST respond with a `404 Not Found` status.
```

Comment:
I think we should force challenge to be used for all requests for a specific file if a Upload-Secret was provided during creation, regardless of the method being used. So currently HEAD, PATCH and POST but might be more in the future. All of these should then return 404 if the Upload-Challenge is invalid. This way, the challenge extension could be used as a very simple user access control.

Reply:
Good idea. I'll change the text accordingly.

Reply:
Update: just pushed this change.